src/cdk/testing/testing.md

```diff
@@ -1 +1,172 @@
+pages. 
+
+The component harness system supports multiple testing environments. You can use the same harness
+implementation in both unit and end-to-end tests. This means that users only need to learn one API,
+and component authors don't have to maintain separate unit and end-to-end test implementations.
+
+Common component libraries, in particular, benefit from this infrastructure due to the wide use of
+their components. Providing a test harness allows the consumers of a component can write tests that
+avoid dependencies on any private implementation details. By capturing these implementation details
+in a single place, consumers can more easily update to new library versions.
+
+This document provides guidance for three types of developers:
+1. [Test authors](#api-for-test-authors)
+2. [Component harness authors](#api-for-component-harness-authors)
+3. [Harness environment authors](#api-for-harness-environment-authors)
+   
+Since many developers fall into only one of these categories, the relevant APIs are broken out by
+developer type in the sections below.
+
+### API for test authors
+
+Test authors are developers using component harnesses written by someone else to test their
+application. For example, this could be an app developer who uses a third-party menu component and
+needs to interact with the menu in a unit test.
+
+#### `ComponentHarness`
+
+This is the abstract base class for all component harnesses. Every harness extends
+`ComponentHarness`. All `ComponentHarness` subclasses have a static property, `hostSelector`, that
+matches the harness class to instances of the component in the DOM. Beyond that, the API of any
+given harness is specific to its corresponding component; refer to the component's documentation to
+learn how to use a specific harness.
+
+#### `TestbedHarnessEnvironment` and `ProtractorHarnessEnvironment`
+
+These classes correspond to different implementations of the component harness system with bindings
+for specific test environments. Any given test must only import _one_ of these classes. Karma-based
+unit tests should use the `TestbedHarnessEnvironment`, while Protractor-based end-to-end tests
+should use the `ProtractorHarnessEnvironment`. Additional environments require custom bindings; see
+[API for harness environment authors](#api-for-harness-environment-authors) for more information on
+alternate test environments.
+
+These classes are primarily used to create a `HarnessLoader` instance, and in certain cases, to
+create `ComponentHarness` instances directly.
+
+`TestbedHarnessEnvironment` offers the following static methods:
+
+| Method | Description |
+| ------ | ----------- |
+| `loader(fixture: ComponentFixture<unknown>): HarnessLoader` | Gets a `HarnessLoader` instance for the given fixture, rooted at the fixture's root element. Should be used to create harnesses for elements contained inside the fixture |
```

Comment:
I'm not sure whether we should be putting API docs in here, because they're bound to go out of sync with the source code. Could we put this in the doc strings in the source instead?

Reply:
Yeah, I'm not sure, a lot of our components docs have a table of @Inputs and @Outputs (and sometimes methods). I think its kind of nice to have it there in context when we're talking about it and showing examples